src/backend/distributed/planner/intermediate_result_pruning.c

```diff
@@ -0,0 +1,293 @@
+	foreach(subPlanCell, subPlanList)
+	{
+		DistributedSubPlan *subPlan = (DistributedSubPlan *) lfirst(subPlanCell);
+		CustomScan *customScan = FetchCitusCustomScanIfExists(subPlan->plan->planTree);
+		if (customScan)
+		{
+			DistributedPlan *distributedPlanOfSubPlan = GetDistributedPlan(customScan);
+			RecordSubplanExecutionsOnNodes(intermediateResultsHash,
+										   distributedPlanOfSubPlan);
+		}
+	}
+}
+
+
+/*
+ * AppendAllAccessedWorkerNodes iterates over all the tasks in a distributed plan
+ * to create the list of worker nodes that can be accessed when this plan is executed.
+ *
+ * If there are multiple placements of a Shard, all of them are considered and
+ * all the workers with placements are appended to the list. This effectively
+ * means that if there is a reference table access in the distributed plan, all
+ * the workers will be in the resulting list.
+ */
+static List *
+AppendAllAccessedWorkerNodes(List *workerNodeList, DistributedPlan *distributedPlan, int
+							 workerNodeCount)
+{
+	List *taskList = distributedPlan->workerJob->taskList;
+	ListCell *taskCell = NULL;
+
+	foreach(taskCell, taskList)
+	{
+		Task *task = lfirst(taskCell);
+		ListCell *placementCell = NULL;
+		foreach(placementCell, task->taskPlacementList)
+		{
+			ShardPlacement *placement = lfirst(placementCell);
+			workerNodeList = list_append_unique_int(workerNodeList, placement->nodeId);
+
+			/* early return if all the workers are accessed */
+			if (list_length(workerNodeList) == workerNodeCount)
+			{
+				return workerNodeList;
+			}
+		}
+	}
+
+	return workerNodeList;
+}
+
+
+/*
+ * MakeIntermediateResultHTAB is a helper method that creates a Hash Table that
+ * stores information on the intermediate result.
+ */
+HTAB *
+MakeIntermediateResultHTAB()
+{
+	HTAB *intermediateResultsHash = NULL;
+	uint32 hashFlags = 0;
+	HASHCTL info = { 0 };
+	int initialNumberOfElements = 16;
+
+	info.keysize = sizeof(IntermediateResultsHashKey);
+	info.entrysize = sizeof(IntermediateResultsHashEntry);
+	info.hash = string_hash;
+	info.hcxt = CurrentMemoryContext;
+	hashFlags = (HASH_ELEM | HASH_FUNCTION | HASH_CONTEXT);
+
+	intermediateResultsHash = hash_create("Intermediate results hash",
+										  initialNumberOfElements, &info, hashFlags);
+
+	return intermediateResultsHash;
+}
+
+
+/*
+ * FindAllWorkerNodesUsingSubplan creates a list of worker nodes that
+ * may need to access subplan results.
+ */
+List *
+FindAllWorkerNodesUsingSubplan(HTAB *intermediateResultsHash,
+							   char *resultId)
+{
+	List *workerNodeList = NIL;
+	IntermediateResultsHashEntry *entry =
+		SearchIntermediateResult(intermediateResultsHash, resultId);
+
+	ListCell *nodeIdCell = NULL;
+	foreach(nodeIdCell, entry->nodeIdList)
+	{
+		WorkerNode *workerNode = LookupNodeByNodeId(lfirst_int(nodeIdCell));
+
+		workerNodeList = lappend(workerNodeList, workerNode);
+		elog(DEBUG1, "Subplan %s will be broadcasted to %s:%d", resultId,
+			 workerNode->workerName, workerNode->workerPort);
```

Comment:
same here, broadcasted->sent and DEBUG5

Reply:
If we do that, we won't have any explicit tests for showing the pruning works correctly. So, I prefer to keep it as `DEBUG1`, not for usefulness for users, but for testing.

Reply:
@marcocitus are you OK with that? The debug message is also not irrelevant, when the log level is DEBUG1, we already print bunch of things for recursive planning.

Reply:
No, let's just change client_min_messages and show it in 3-4 tests.

Reply:
I created a hidden GUC `log_intermediate_results` that handles the debug outputs for this case.

By default the messages will be shown if `client_min_messages` is `DEBUG4` or below. If the GUC is set, they will be shown for `DEBUG1` and below.